### PR TITLE
Minimatch for directory exclusion

### PR DESCRIPTION
**Pull request by chriskarlsson**
To me the description on how to define DirectoryExclusionList was very confusing since the comparison was globing and not minimatch and that there was no mention of ';' being used to define multiple patterns. This PR aims to solve both of those issues.

The minimatch implementation is a port of https://github.com/isaacs/minimatch to C#.

### CODE REVIEW

**Comment by chriskarlsson** [comment]
@microsoft-github-policy-service agree company="Appgate"

File: src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs [hunk 1]
```diff
@@ -0,0 +1,843 @@
+﻿namespace Microsoft.ComponentDetection.Orchestrator;
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+/// <summary>
```
[the remark this answers is not in the document]
Comment (cobya) [reply src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs:8]: Assuming your reference was latest, this would be from commit [cb4be48a55d64b3a40a745d4a8eb4d1b06507277](https://github.com/isaacs/minimatch/commit/cb4be48a55d64b3a40a745d4a8eb4d1b06507277)

[the remark this answers is not in the document]
Comment (cobya) [reply src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs:8]: Let's also make sure to follow the terms of the [license agreement](https://github.com/isaacs/minimatch/blob/main/LICENSE) as well. This essentially means copying the license into this file.

File: src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs [hunk 2]
```diff
@@ -0,0 +1,843 @@
...
+            // ran out of pattern, still have file left.
+            // this is only acceptable if we're on the very last
+            // empty segment of a file with a trailing slash.
+            // a/* should match a/b/
+            var emptyFileEnd = segmentIndex == segments.Count - 1 && segments[segmentIndex].Length == 0;
+            return emptyFileEnd;
+        }
+
+        // should be unreachable.
+        throw new InvalidOperationException("wtf?");
```
Comment (cobya) [inline src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs:790]: nit: I know it comes from the original source, but can we give this a message that is more clear?

File: test/Microsoft.ComponentDetection.Orchestrator.Tests/MinimatchTests.cs [hunk 3]
```diff
@@ -0,0 +1,197 @@
+namespace Microsoft.ComponentDetection.Orchestrator.Tests;
+
+using System;
+using System.Linq;
+using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+[TestClass]
```
Comment (cobya) [inline test/Microsoft.ComponentDetection.Orchestrator.Tests/MinimatchTests.cs:8]: Thank you for including test coverage 😄 If you're up for it, I think we could get a pretty solid test coverage percentage by using some of the tests below:
- https://github.com/isaacs/minimatch/blob/main/test/basic.js
- https://github.com/isaacs/minimatch/blob/main/test/class-edge-cases.js
- https://github.com/isaacs/minimatch/blob/main/test/tricky-negations.js

**Review by cobya — CHANGES_REQUESTED** [review]
First of all, thank you for your contribution to the repo! There are some things that I've noted which would be needed before we're able to accept the contribution 👍 

Since this would be a breaking change in how our directory exclusion pattern works for downstream consumers, I've tagged it for the next major release of Component Detection.

Do you have a timeline when this is needed for your tooling? Generally speaking we try and give a few weeks heads before pushing major versions.

**Comment by chriskarlsson** [comment]
Sorry. I've been quite busy the last weeks. I hope I could get back to it this week.

File: src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (chriskarlsson) [reply src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs:8]: Will fix.

File: src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs [hunk 4]
```diff
@@ -0,0 +1,843 @@
...
+    // don't need to escape / when using new RegExp()
+    private const string Qmark = "[^/]";
+
+    // any single thing other than /
+    // don't need to escape / when using new RegExp()
+    private const string Star = Qmark + "*?";
+
+    // characters that need to be escaped in RegExp.
+    private static readonly HashSet<char> ReSpecials = new("().*{}+?[]^$\\!".ToCharArray());
+    private static readonly Regex SlashSplit = new("/+");
```
[the remark this answers is not in the document]
Comment (chriskarlsson) [reply src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs:23]: Sounds reasonable.

**Comment by chriskarlsson** [comment]
> Do you have a timeline when this is needed for your tooling? Generally speaking we try and give a few weeks heads before pushing major versions.

Missed your question before. I've worked around this in our tooling. So for us this isn't necessary. Just thought it would be a nice upgrade to the repo.

**Comment by chriskarlsson** [comment]
@cobya - Updated with respect to your comments last week. Please have a look to see if it looks good to you.

File: test/Microsoft.ComponentDetection.Orchestrator.Tests/MinimatchTests.cs (hunk 3 above)
Comment (cobya) [reply test/Microsoft.ComponentDetection.Orchestrator.Tests/MinimatchTests.cs:10]: Thanks for the additional tests!

File: src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cobya) [reply src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs:30]: Thanks for adding this in 😃

**Comment by cobya** [comment]
Thanks again for the contribution here 🫶 I've merged the changes into the `cobya/v5` branch will be used as the baseline for the upcoming breaking changes tracked in #1135 and the next major release!
